lib/indexer/addrindexer.js

```diff
@@ -0,0 +1,259 @@
+
+  fromRaw(data) {
+    const br = bio.read(data);
+
+    this.height = br.readU32();
+    this.index = br.readU32();
+
+    return this;
+  }
+
+  /**
+   * Instantiate a count from a buffer.
+   * @param {Buffer} data
+   * @returns {Count}
+   */
+
+  static fromRaw(data) {
+    return new this().fromRaw(data);
+  }
+}
+
+/**
+ * AddrIndexer
+ * @alias module:indexer.AddrIndexer
+ * @extends Indexer
+ */
+
+class AddrIndexer extends Indexer {
+  /**
+   * Create a indexer
+   * @constructor
+   * @param {Object} options
+   */
+
+  constructor(options) {
+    super('addr', options);
+
+    this.db = bdb.create(this.options);
+    this.maxTxs = options.maxTxs || 100;
+  }
+
+  /**
+   * Index transactions by address.
+   * @private
+   * @param {ChainEntry} entry
+   * @param {Block} block
+   * @param {CoinView} view
+   */
+
+  async indexBlock(entry, block, view) {
+    const height = entry.height;
+
+    for (let i = 0; i < block.txs.length; i++) {
+      const tx = block.txs[i];
+      const hash = tx.hash();
+      const count = new Count(height, i);
+
+      this.put(layout.C.encode(height, i), hash);
+      this.put(layout.c.encode(hash), count.toRaw());
+
+      for (const addr of tx.getAddresses(view)) {
+        const prefix = addr.getPrefix();
+
+        if (prefix < 0)
+          continue;
+
+        const addrHash = addr.getHash();
+
+        this.put(layout.A.encode(prefix, addrHash, height, i), null);
+      }
+    }
+  }
+
+  /**
+   * Remove addresses from index.
+   * @private
+   * @param {ChainEntry} entry
+   * @param {Block} block
+   * @param {CoinView} view
+   */
+
+  async unindexBlock(entry, block, view) {
+    const height = entry.height;
+
+    for (let i = 0; i < block.txs.length; i++) {
+      const tx = block.txs[i];
+      const hash = tx.hash();
+
+      this.del(layout.C.encode(height, i));
+      this.del(layout.c.encode(hash));
+
+      for (const addr of tx.getAddresses(view)) {
+        const prefix = addr.getPrefix();
+
+        if (prefix < 0)
+          continue;
+
+        const addrHash = addr.getHash();
+
+        this.del(layout.A.encode(prefix, addrHash, height, i));
+      }
+    }
+  }
+
+  /**
+   * Get transaction hashes to an address in ascending or descending
+   * order. If the `after` argument is supplied, results will be given
+   * _after_ that transaction hash. The default order is ascending from
+   * oldest to latest.
+   * @param {Address} addr
+   * @param {Object} options
+   * @param {Buffer} options.after - A transaction hash
+   * @param {Number} options.limit
+   * @param {Boolean} options.reverse
+   * @returns {Promise} - Returns {@link Hash}[].
+   */
+
+  async getHashesByAddress(addr, options = {}) {
+    const txs = [];
+
+    const {after, reverse} = options;
+    let {limit} = options;
+
+    if (!limit)
+      limit = this.maxTxs;
+
+    if (limit > this.maxTxs)
+      throw new Error('Limit above max of ${this.maxTxs}.');
+
+    const hash = Address.getHash(addr);
```

Comment:
Neither `hash` or `prefix` are used unless `after && hasAfter`. These function invocations could be moved into that code path. I also think that it wouldn't hurt to add some comments in this function to explain some of the logic, particularly around `hasAfter` and `skip`

Reply:
That is not correct, the address `prefix` and `hash` are used for every query. There is test coverage around `hasAfter` and `skip`, it relates to combining the results of unconfirmed and confirmed transactions. If this is an `after` query, the `after` hash is not found, and this confirmed query is before the unconfirmed query (e.g. `!reverse`),  the `after` hash is expected to be within the unconfirmed query, so skip all results.